### ISSUE TITLE
ECDSA wallet DKG

### ISSUE THREAD

**Issue by pdyraga** [keep-network/keep-core#2758]
Depends on #2756 

Important: This issue is only about DKG, signature requests are out of scope.

For tBTC v2, we need tBTC bridge to create ECDSA wallets in the network. We should use the same, submit-challenge/approve approach for DKG as in random beacon v2 except that instead of a BLS key, the group should submit an ECDSA public key. We will start in `solidity/ecdsa` with `WalletFactory` and `Wallet` contracts. We should be able to import and reuse random beacon `DKG` and `DKGValidator` parametrizing group size, group threshold, and active threshold. I strongly suggest we do not try to make group size and threshold parameters of the create wallet function but hardcode them in the wallet as constants and assume a change of the group size is a hard-fork event of tBTC bridge component.

The total gas cost of DKG (submit+approve) is of paramount importance and after the changes from #2756 it should not exceed 669k. For a wallet created once a week, this is still a lot and should be optimized aggressively when working on this code. For example, we need to validate what is the difference between cloning `Wallet` contract and using a single `Wallet` contract dispatching signing requests by virtual wallet ID, in a similar way how signing requests are dispatched to groups in `RandomBeacon`.

Reusing `Groups` library from the random beacon is probably not a good idea at this point given that ECDSA wallets will have completely other lifecycle properties than beacon groups and failing heartbeat means something completely else. If we end up duplicating `Groups` library code now (with no expiration/termination/selection), this is fine.

Wallet API should be designed as a part of the work on this issue, two example options are presented below. The first one feels way more elegant but we need to take into account gas costs - to be decided when working on this issue.

```
contract WalletFactory {
  function createNewWallet(): Wallet
}

contract Wallet {
  function submitDkgResult(DKG.Result)
  function approveDkgResult(DKG.Result)
  function challengeDkgResult(DKG.Result)

  // later:
  function sign(bytes32 digest)
}
```

```
contract Wallet {
  function createNew(): uint // wallet ID
  function submitDKGResult(uint walletID, DKG.Result);
  function approveDkgResult(uint walletID, DKG.Result)
  function challengeDkgResult(uint walletID, DKG.Result)

  // later:
  function sign(uint walletID, bytes32 digest)
}
```

**Comment by nkuba** [comment]
What if we link `DKG` library to the `WalletFactory` and create a wallet after DKG completes (inside `approveDkgResult` function or create Wallet in `submitDkgResult` and just change the Wallet state in `approveDkgResult`))? We don't expect two DKGs to happen at the same time, as we will lock the sortition pool, right? Operators would have to register and interact with one `WalletFactory` contract and only after DKG is completed they will subscribe to the individual `Wallet` events.

```
contract WalletFactory {
  function createNewWallet(): Wallet

  function submitDkgResult(DKG.Result)
  function approveDkgResult(DKG.Result)
  function challengeDkgResult(DKG.Result)

}

contract Wallet {
  // later:
  function sign(bytes32 digest)
}

```

**Comment by nkuba** [comment]
I compared the gas cost of three different approaches:

### 1. Wallets registered in a mapping in a single library (similar to [Groups in Random Beacon V2](https://github.com/keep-network/keep-core/blob/6a3be3a5abe2d239f1e986ce24f051d08bc55c9d/solidity/random-beacon/contracts/libraries/Groups.sol))

```
·------------------------------------------------|---------------------------|-------------|-----------------------------·
|              Solc version: 0.8.9               ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 30000000 gas  │
·················································|···························|·············|······························
|  Methods                                                                                                               │
·······················|·························|·············|·············|·············|·············|················
|  Contract            ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls    ·  eur (avg)    │
·······················|·························|·············|·············|·············|·············|················
|  StakingStub         ·  increaseAuthorization  ·      44858  ·      44870  ·      44869  ·         66  ·            -  │
·······················|·························|·············|·············|·············|·············|················
|  WalletRegistryStub  ·  approveDkgResult       ·          -  ·          -  ·     267881  ·          2  ·            -  │
·······················|·························|·············|·············|·············|·············|················
|  WalletRegistryStub  ·  registerOperator       ·     205970  ·     342709  ·     210730  ·         66  ·            -  │
·······················|·························|·············|·············|·············|·············|················
|  WalletRegistryStub  ·  requestNewWallet       ·          -  ·          -  ·      82574  ·          1  ·            -  │
·······················|·························|·············|·············|·············|·············|················
|  WalletRegistryStub  ·  sign                   ·          -  ·          -  ·      47762  ·          1  ·            -  │
·······················|·························|·············|·············|·············|·············|················
|  WalletRegistryStub  ·  submitDkgResult        ·          -  ·          -  ·     300390  ·          1  ·            -  │
·······················|·························|·············|·············|·············|·············|················
|  WalletRegistryStub  ·  submitSignature        ·          -  ·          -  ·      35942  ·          1  ·            -  │
·······················|·························|·············|·············|·············|·············|················
|  WalletRegistryStub  ·  updateDkgParams        ·          -  ·          -  ·      40579  ·          1  ·            -  │
·----------------------|-------------------------|-------------|-------------|-------------|-------------|---------------·
```

### 2. Wallets deployed as separate contract instances with Clone Factory (`createClone(masterWallet)` - similar to [BondedECDSAKeep creation in Keep ECDSA V1](https://github.com/keep-network/keep-ecdsa/blob/49b83e4471c64879d990a572950cf9b9a0d30468/solidity/contracts/KeepCreator.sol))

```
·--------------------------------------------|---------------------------|-------------|-----------------------------·
|            Solc version: 0.8.9             ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 30000000 gas  │
·············································|···························|·············|······························
|  Methods                                                                                                           │
···················|·························|·············|·············|·············|·············|················
|  Contract        ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls    ·  eur (avg)    │
···················|·························|·············|·············|·············|·············|················
|  StakingStub     ·  increaseAuthorization  ·      44858  ·      44870  ·      44869  ·         66  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  Wallet          ·  sign                   ·          -  ·          -  ·      49877  ·          1  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  Wallet          ·  submitSignature        ·          -  ·          -  ·      34417  ·          1  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  approveDkgResult       ·          -  ·          -  ·     314506  ·          2  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  registerOperator       ·     205970  ·     342709  ·     210730  ·         66  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  requestNewWallet       ·          -  ·          -  ·      82552  ·          1  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  submitDkgResult        ·          -  ·          -  ·     300306  ·          1  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  updateDkgParams        ·          -  ·          -  ·      40579  ·          1  ·            -  │
·------------------|-------------------------|-------------|-------------|-------------|-------------|---------------·
```

### 3. Wallets created as separate contract instances (`new Wallet()`)

```
·--------------------------------------------|---------------------------|-------------|-----------------------------·
|            Solc version: 0.8.9             ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 30000000 gas  │
·············································|···························|·············|······························
|  Methods                                                                                                           │
···················|·························|·············|·············|·············|·············|················
|  Contract        ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls    ·  eur (avg)    │
···················|·························|·············|·············|·············|·············|················
|  StakingStub     ·  increaseAuthorization  ·      44858  ·      44870  ·      44869  ·         66  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  Wallet          ·  sign                   ·          -  ·          -  ·      47208  ·          1  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  Wallet          ·  submitSignature        ·          -  ·          -  ·      31734  ·          1  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  approveDkgResult       ·          -  ·          -  ·     667853  ·          2  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  registerOperator       ·     205970  ·     342709  ·     210730  ·         66  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  requestNewWallet       ·          -  ·          -  ·      82552  ·          1  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  submitDkgResult        ·          -  ·          -  ·     300306  ·          1  ·            -  │
···················|·························|·············|·············|·············|·············|················
|  WalletRegistry  ·  updateDkgParams        ·          -  ·          -  ·      40579  ·          1  ·            -  │
·------------------|-------------------------|-------------|-------------|-------------|-------------|---------------·
```


NOTE: Those are not the final costs as the function had just an initial stub implementation.
